server/apis/v1/handler.go

```diff
@@ -1180,6 +1185,123 @@ func (h *handler) GetMonoVertexHealth(c *gin.Context) {
+func validateMetricReqBody(data MetricSpecData) error {
+	if (data.From == "" && data.To != "") || (data.To == "" && data.From != "") {
+		return fmt.Errorf("bad request: either both from and to be set or none")
+	}
+
+	if _, ok := metricNameMap[data.MetricName]; !ok {
+		return fmt.Errorf("bad Request: requested user metric name not supported yet")
+	}
+	return nil
+}
+
+// build query from labels provided by user
+func buildMetricLabelQuery(data map[string]string) string {
+	query := ""
+	for key, val := range data {
+		query += fmt.Sprintf(`,%s="%s"`, key, val)
+	}
+	return query
+}
+
+// build time series data from prometheus response which can be in Matrix or Vector format
+func buildTimeSeriesData(ctx context.Context, v1api v1.API, data MetricSpecData, query string, format PrometheusResponseFormat) ([][]float64, error) {
+	var timeSeriesData = make([][]float64, 0)
+	if format == Matrix {
+		startTime, _ := time.Parse(time.RFC3339, data.From)
+		endTime, _ := time.Parse(time.RFC3339, data.To)
+		r := v1.Range{
+			Start: startTime,
+			End:   endTime,
+			Step:  time.Minute,
+		}
+		result, _, err := v1api.QueryRange(ctx, query, r)
+		if err != nil {
+			return timeSeriesData, err
+		}
+		matrixRes := result.(model.Matrix)
+
+		for _, sampleStream := range matrixRes {
+			for _, val := range sampleStream.Values {
+				timeSeriesData = append(timeSeriesData, []float64{
+					float64(val.Timestamp),
+					float64(val.Value),
+				})
+			}
+		}
+
+	} else if format == Vector {
+		result, _, err := v1api.Query(ctx, query, model.Now().Time())
+		if err != nil {
+			return timeSeriesData, err
+		}
+		vectorRes := result.(model.Vector)
+		for _, sample := range vectorRes {
+			timeSeriesData = append(timeSeriesData, []float64{
+				float64(sample.Timestamp),
+				float64(sample.Value),
+			})
+		}
+	}
+	return timeSeriesData, nil
+}
+
+func (h *handler) GetMetricData(c *gin.Context) {
+
+	if h.prometheusServerUrl == "" {
+		h.respondWithError(c, "Prometheus Server URL not set.")
+		return
+	}
+
+	ns, pipeline := c.Param("namespace"), c.Param("pipeline")
+
+	// bind request body to metric specs data
+	var metricSpecData MetricSpecData
+	if err := bindJson(c, &metricSpecData); err != nil {
+		h.respondWithError(c, fmt.Sprintf("Failed to decode JSON request body to metric data spec, %s", err.Error()))
+		return
+	}
+	if err := validateMetricReqBody(metricSpecData); err != nil {
+		h.respondWithError(c, fmt.Sprintf("Validation check failed for metric data body, %s", err.Error()))
+		return
+	}
+
+	// build labels query
+	labelQuery := buildMetricLabelQuery(metricSpecData.Labels)
+
+	// query builder: "rate" hardcoded for counter metrics - first iteration - revisit later
+	query := fmt.Sprintf(`rate(%s{namespace="%s",pipeline="%s"%s}[%s])`, metricNameMap[metricSpecData.MetricName].NumaMetricName, ns, pipeline, labelQuery, metricSpecData.Duration)
+
+	client, err := api.NewClient(api.Config{
+		Address: h.prometheusServerUrl,
+	})
```

Comment:
We will end up creating a new client for every api invocation. Can we store it inside the handler struct and reuse it?

Reply:
yes, will resolve this.